### PR TITLE
#5050 Remember the position and size of the speechViewer

### LINKED ISSUE #5050: Option to show Speech Viewer by default (and save dimensions and position of the window)
**Reported by jmuheim on 2015-04-23 06:41**
I'm a sighted user and I'm using NVDA only with "no speech" option and with the speech viewer open. This makes it much faster for me to use, as I can jump to the relevant parts of the output directly by scanning through the speech viewer visually.

I know that I can open the speech viewer using NVDA+N, then T, then S, but this is cumbersome. I'd like to have an option "Display speech viewer when starting NVDA", as I normally start NVDA only for a few seconds, and then I quit it again so it doesn't interfere my keyboard (and slow down my computer, which is a Windows 7 virtual machine on Mac OS X).

In addition, it would be very nice if the window's last dimensions (width/height) and position (x/y) were remembered, as I usually want it side by side with my browser so they don't overlap each other.

I really hope this will be implemented. I'm working for the best known accessibility consultancy in Switzerland, and we are trying to get our clients to use NVDA themselves to produce highly accessible websites themselves without our help. So it would really make sense to make NVDA appeal a little bit more to sighted users, as typically it's them who create the websites that blind people use.

### PR DESCRIPTION
**Pull request by feerrenrut**
Partially resolves #5050 (remember the position and size of the speech viewer window)

The speechViewer window will open in the same location at the same size as it was when previously closed. If the monitor setup has changed (resolution, or number of monitors) the window is auto positioned/sized.

### CODE REVIEW

**Comment by feerrenrut** [comment]
@michaelDCurran please take a look when you get a chance.

**Comment by feerrenrut** [comment]
- [ ] Test saving and restoring config profiles

File: source/speechViewer.py [hunk 1]
```diff
@@ -27,6 +36,31 @@ def onClose(self, evt):
...
+
+	def savePositionInformation(self):
+		position = self.GetPosition()
+		config.conf["speechView"]["x"] = position.x
+		config.conf["speechView"]["y"] = position.y
+		size = self.GetSize()
+		config.conf["speechView"]["width"] = size.width
+		config.conf["speechView"]["height"] = size.height
+		config.conf["speechView"]["displays"] = self.getAttachedDisplaySizes()
+		config.conf["speechView"]["autoPositionWindow"] = False
```
Comment (feerrenrut) [inline source/speechViewer.py]: As discussed with @michaelDCurran add these to the config schema. Perhaps the displays entry could just hold the hash of the object. Though now I'm thinking about this again, using the hash would have the disadvantage that it becomes much harder to verify the config / behaviour.

**Comment by michaelDCurran** [comment]
Change range to xrange, but other than that this looks fine.

**Comment by jmuheim** [comment]
Hello

My speech viewer is still not remembering its position after a relaunch of NVDA. Do I misunderstand something? Do I need to check an option to make this work?

**Comment by feerrenrut** [comment]
>If the monitor setup has changed (resolution, or number of monitors) the window is auto positioned/sized.

As long as there is no change in displays and the config is saved / not destroyed, it should save and reuse the location.

Do you have "Save configuration when exiting NVDA" enabled?
If not, the config (including position information) may not have been saved, this type of preference could potentially be independent of that setting. It's not really user config, instead it is preserved application state.

**Comment by jmuheim** [comment]
Suddenly, since I wrote the comment above, it seems to work! Very interesting... Didn't work before, and I don't think that I changed anything in the settings.

Whatever! It's working now, that's all that counts. Thank you!

**Pull request merged**